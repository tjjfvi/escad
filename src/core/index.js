
const builtins = require("./builtins");

module.exports = { ...builtins, ...builtins.operators };
