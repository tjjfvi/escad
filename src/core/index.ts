
export * from "./builtins";
import * as builtins from "./builtins";
import { Elementish } from "./Element";

export default Object.assign(<T extends builtins.Product>(...a: Elementish<T>[]) =>
  new builtins.Element(a.length === 1 ? a[0] : a), builtins);
