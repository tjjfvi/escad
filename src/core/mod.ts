import {
  defaultChainables,
  extendChainables,
  pickChainables,
  Realm,
} from "./chaining/mod.ts";

import * as _chainables from "./chainables/mod.ts";
const chainables = pickChainables(_chainables);
export default chainables;
type Chainables = typeof chainables;
declare global {
  export namespace escad {
    export interface DefaultChainables extends Chainables {}
  }
}
extendChainables(chainables);

export const escad = Realm.create(() => defaultChainables);

// moderate

export * from "./artifacts/mod.ts";
export * from "./chainables/mod.ts";
export * from "./chaining/mod.ts";
export * from "./conversions/mod.ts";
export * from "./exports/mod.ts";
export * from "./hierarchy/mod.ts";
export * from "./logs/mod.ts";
export * from "./parameters/mod.ts";
export * from "./product/mod.ts";
export * from "./serializers/mod.ts";
export * from "./utils/mod.ts";
