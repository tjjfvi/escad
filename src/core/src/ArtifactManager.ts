
import { WeakCache } from "./WeakCache"
import { Hash } from "./Hash"
import { ArtifactStore } from "./ArtifactStore"
import { WrappedValue } from "./WrappedValue"
import { Counter, Timer } from "./Timer"

export class ArtifactManager {

  private excludeStoresCaches = new WeakMap<ReadonlySet<ArtifactStore>, WeakCache<Hash<unknown>, WrappedValue | null>>()
  private defaultCache = new WeakCache<string, WrappedValue | null>()

  artifactStores: ArtifactStore[] = []

  timers = Timer.create({
    storeRaw: Timer.create(),
    storeRef: Timer.create(),
    lookupRaw: Timer.create({ calls: Counter.create() }),
    lookupRef: Timer.create({ calls: Counter.create() }),
    getRefCacheKey: Timer.create(),
  })

  async storeRaw<T>(
    artifactPromise: T | Promise<T>,
    excludeStores?: ReadonlySet<ArtifactStore>,
  ){
    const artifact = await artifactPromise

    this.timers.storeRaw.start()
    const artifactHash = Hash.create(artifact)

    this.getCache(excludeStores).set(artifactHash, WrappedValue.create(artifact))

    await Promise.all(this.artifactStores.map(async s => {
      if(!excludeStores?.has(s))
        await s.storeRaw?.(artifactHash, WrappedValue.create(artifact), this)
    }))

    this.timers.storeRaw.end()
    return artifactHash
  }

  async storeRef<T>(
    loc: readonly unknown[],
    artifactPromise: T | Promise<T>,
    excludeStores?: ReadonlySet<ArtifactStore>,
  ){
    this.getCache(excludeStores).setAsync(
      this.getRefCacheKey(loc),
      Promise.resolve(WrappedValue.create(artifactPromise)),
    )
    const artifact = await artifactPromise

    this.timers.storeRef.start()
    const artifactHash = Hash.create(artifact)

    await Promise.all<unknown>([
      this.storeRaw(artifact, excludeStores),
      ...loc.map(l => this.storeRaw(l, excludeStores)),
      ...this.artifactStores.map(s => {
        if(!excludeStores?.has(s))
          s.storeRef?.(loc, artifactHash, this)
      }),
    ])

    this.timers.storeRef.end()
    return artifactHash
  }

  async lookupRaw<T>(
    hash: Hash<T>,
    excludeStores?: ReadonlySet<ArtifactStore>,
  ){
    const result = await this.lookupRawWrapped(hash, excludeStores)
    if(result) return result.value
    return null
  }

  lookupRawWrapped<T>(
    hash: Hash<T>,
    excludeStores?: ReadonlySet<ArtifactStore>,
  ){
    this.timers.lookupRaw.calls.increment()
    return this.getCache(excludeStores).getAsync(
      hash,
      this.timers.lookupRaw.timeFn(async (): Promise<WrappedValue<T> | null> => {
        for(const store of this.artifactStores)
          if(!excludeStores?.has(store)) {
            const artifact = await store.lookupRaw?.(hash, this)
            if(artifact && Hash.check(hash, artifact.value))
              return artifact as WrappedValue<T>
          }
        return null
      }),
    ) as Promise<WrappedValue<T> | null>
  }

  async lookupRef(
    loc: readonly unknown[],
    excludeStores?: ReadonlySet<ArtifactStore>,
  ){
    const result = await this.lookupRefWrapped(loc, excludeStores)
    if(result) return result.value
    return null
  }

  lookupRefWrapped(
    loc: readonly unknown[],
    excludeStores?: ReadonlySet<ArtifactStore>,
  ): Promise<WrappedValue | null>{
    this.timers.lookupRef.calls.increment()
    return this.getCache(excludeStores).getAsync(
      this.getRefCacheKey(loc),
      this.timers.lookupRef.timeFn(async () => {
        for(const store of this.artifactStores)
          if(!excludeStores?.has(store)) {
            const artifact = await store.lookupRef?.(loc, this)
            if(artifact) {
              await this.storeRaw(artifact.value, excludeStores)
              return artifact
            }
          }
        return null
      }),
    )
  }

  async computeRef<T>(
    loc: readonly unknown[],
    func: () => Promise<T>,
    excludeStores?: ReadonlySet<ArtifactStore>,
  ){
    const stored = await this.lookupRefWrapped(loc, excludeStores)
    if(stored) return stored.value as T
    const result = func()
    this.storeRef(loc, result, excludeStores)
    return await result
  }

  private getCache(excludeStores?: ReadonlySet<ArtifactStore>){
    if(!excludeStores?.size)
      return this.defaultCache
    const existing = this.excludeStoresCaches.get(excludeStores)
    if(existing) return existing
    const cache = new WeakCache<Hash<unknown>, WrappedValue>()
    this.excludeStoresCaches.set(excludeStores, cache)
    return cache
  }

  private getRefCacheKey(loc: readonly unknown[]){
    this.timers.getRefCacheKey.start()
    let key = ""
    for(const part of loc)
      if(typeof part === "string")
        key += part + "//"
      else
        key += Hash.create(part) + "//"
    this.timers.getRefCacheKey.end()
    return key
  }

}

export const artifactManager = new ArtifactManager()
