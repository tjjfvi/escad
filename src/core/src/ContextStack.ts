
import { Id } from "./Id"

export class ContextStack {

  private root: Record<string, unknown> = Object.create(null)
  private contextMap = this.root

  wrap<U>(fn: () => U): U{
    const orig = this.contextMap
    this.contextMap = Object.create(this.contextMap)
    const result = fn()
    this.contextMap = orig
    return result
  }

  get(id: Id): unknown{
    return this.contextMap[id]
  }

  set(id: Id, value: unknown){
    this.contextMap[id] = value
  }

}

export const contextStack = new ContextStack()
