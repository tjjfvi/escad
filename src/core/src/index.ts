
export * from "./defaultChainables"
import { defaultChainables } from "./defaultChainables"
import { Realm } from "./Realm"

export const escad = Realm.create(() => defaultChainables)
export default escad

// @create-index {"mode":"*"}

export * from "./ArrayHierarchy"
export * from "./ArrayProduct"
export * from "./ArtifactManager"
export * from "./ArtifactStore"
export * from "./CallHierarchy"
export * from "./Component"
export * from "./Context"
export * from "./ContextStack"
export * from "./ConversionRegistry"
export * from "./Conversions"
export * from "./Element"
export * from "./ExportType"
export * from "./ExportTypeRegistry"
export * from "./ExtensibleFunction"
export * from "./Hash"
export * from "./HashMap"
export * from "./HashProduct"
export * from "./HashSet"
export * from "./Hierarchy"
export * from "./HierarchyLog"
export * from "./Hkt"
export * from "./Id"
export * from "./InMemoryArtifactStore"
export * from "./LeafProduct"
export * from "./Log"
export * from "./Logger"
export * from "./MarkedProduct"
export * from "./MultiHashMap"
export * from "./NameHierarchy"
export * from "./ObjectHierarchy"
export * from "./ObjectParam"
export * from "./Operation"
export * from "./Parameter"
export * from "./Product"
export * from "./Promisish"
export * from "./Realm"
export * from "./RealmComponent"
export * from "./RealmElement"
export * from "./RealmOperation"
export * from "./RealmThing"
export * from "./StringLog"
export * from "./Thing"
export * from "./Timer"
export * from "./TupleProduct"
export * from "./UnknownProduct"
export * from "./ValueHierarchy"
export * from "./WeakCache"
export * from "./WrappedValue"
export * from "./assertNever"
export * from "./checkTypeProperty"
export * from "./defaultChainables"
export * from "./depthFirst"
export * from "./mapOperation"
export * from "./serializers"

