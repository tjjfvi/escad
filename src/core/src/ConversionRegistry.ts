/* eslint-disable @typescript-eslint/no-non-null-assertion */

import { ProductType, Product } from "./Product"
import { ConvertibleTo, ConversionImpl, ConversionImplish } from "./Conversions"
import { Hash } from "./Hash"
import { artifactManager, ArtifactManager } from "./ArtifactManager"
import { ArtifactStore } from "./ArtifactStore"
import { Id } from "./Id"
import { TupleProduct, TupleProductType } from "./TupleProduct"
import { ArrayProduct, ArrayProductType } from "./ArrayProduct"
import { UnknownProduct, UnknownProductType } from "./UnknownProduct"
import { MarkedProduct, MarkedProductType } from "./MarkedProduct"
import { HashProduct, HashProductType } from "./HashProduct"
import { WrappedValue } from "./WrappedValue"

type ConversionPath = ConversionImpl<any, any>[]

export class ConversionRegistry {

  constructor(public artifactManager: ArtifactManager){
    this.artifactManager.artifactStores.push(this.artifactStore)
  }

  static readonly artifactStoreId = Id.create(__filename, "@escad/core", "ArtifactStore", "ConversionRegistry")
  readonly artifactStore: ArtifactStore = {
    lookupRef: async ([id, toType, from]) => {
      if(id !== ConversionRegistry.artifactStoreId) return null
      if(!ProductType.isProductType(toType)) return null
      if(Product.isProduct(from))
        return WrappedValue.create(this.convertProduct(toType, from))
      if(ProductType.isProductType(from))
        return WrappedValue.create(this.compose(from, toType))
      return null
    },
  }

  readonly excludeStores: ReadonlySet<ArtifactStore> = new Set([this.artifactStore])

  private readonly registered = new Map<Id, ConversionImpl<any, any>>()

  register<F extends Product, T extends Product>(conversion: ConversionImplish<F, T>): void{
    this.registered.set(conversion.id, {
      ...conversion,
      fromType: ProductType.fromProductTypeish(conversion.fromType),
      toType: ProductType.fromProductTypeish(conversion.toType),
    })
  }

  async has(a: ProductType, b: ProductType){
    return !!(await this.compose(a, b))
  }

  private async compose(fromType: ProductType, toType: ProductType): Promise<ConversionPath | null>{
    const result = await this.artifactManager.computeRef(
      [ConversionRegistry.artifactStoreId, toType, fromType],
      async () => {
        let bestPath = null as ConversionPath | null
        const promises = []

        const tasks = [{
          type: fromType,
          prior: [] as ConversionPath,
        }]

        while(tasks.length) {
          const { type, prior } = tasks.pop()!
          if(prior.some(c => Hash.equal(c.fromType, type)))
            continue

          if(this.maybeImplicitlyConvertibleTo(type, toType))
            promises.push(
              this.finishPath(fromType, prior, toType).then(path => {
                if(!bestPath || path && this.weight(path) <= this.weight(bestPath))
                  bestPath = path
              }),
            )

          else
            for(const conversion of this.registered.values())
              if(this.maybeImplicitlyConvertibleTo(type, conversion.fromType))
                tasks.push({
                  type: conversion.toType,
                  prior: [...prior, conversion],
                })
        }

        await Promise.all(promises)

        return bestPath
      },
      this.excludeStores,
    )

    if(!result) return null

    return this.rehydrateConversionPath(result)
  }

  private async finishPath(initialFromType: ProductType, path: ConversionPath | null, finalToType: ProductType){
    if(!path)
      return null

    for(let i = 0; i <= path.length; i++) {
      const fromType = path[i - 1]?.toType ?? initialFromType
      const toType = path[i]?.fromType ?? finalToType

      if(Hash.equal(fromType, toType))
        continue

      const part = await this.finishPathSegment(fromType, toType)
      if(!part) return null

      path.splice(i, 0, part)
      i--
    }

    return path
  }

  private maybeImplicitlyConvertibleTo(fromType: ProductType, toType: ProductType): boolean{
    return false
      || Hash.equal(fromType, toType)
      || HashProductType.isHashProductType(fromType)
      || HashProductType.isHashProductType(toType)
      || UnknownProductType.isUnknownProductType(toType)
      || true
        && MarkedProductType.isMarkedProductType(fromType)
        && MarkedProductType.isMarkedProductType(toType)
        && fromType.marker === toType.marker
      || true
        && TupleProductType.isTupleProductType(fromType)
        && ArrayProductType.isArrayProductType(toType)
      || true
        && ArrayProductType.isArrayProductType(fromType)
        && ArrayProductType.isArrayProductType(toType)
      || true
        && TupleProductType.isTupleProductType(fromType)
        && TupleProductType.isTupleProductType(toType)
        && fromType.elementTypes.length === toType.elementTypes.length
  }

  private async finishPathSegment(
    fromType: ProductType,
    toType: ProductType,
  ): Promise<Omit<ConversionImpl<any, any>, "id"> | null>{
    if(HashProductType.isHashProductType(fromType)) {
      const resolvedFromType = await this.artifactManager.lookupRaw(fromType.productType)
      if(!resolvedFromType) throw new Error("Could not resolve HashProductType")
      const subPath = await this.compose(resolvedFromType, toType)
      return subPath && {
        fromType,
        toType,
        convert: async ({ hash }: HashProduct) => {
          const product = await this.artifactManager.lookupRaw(hash)
          if(!product)
            throw new Error("Could not resolve HashProduct " + hash)
          return await this.executeConversionPath(product, subPath)
        },
        weight: this.weight(subPath),
      }
    }

    if(HashProductType.isHashProductType(toType)) {
      const resolvedToType = await this.artifactManager.lookupRaw(toType.productType)
      if(!resolvedToType) throw new Error("Could not resolve HashProductType")
      const subPath = await this.compose(fromType, resolvedToType)
      return subPath && {
        fromType: resolvedToType,
        toType,
        convert: async (product: Product) =>
          HashProduct.fromProduct(
            await this.executeConversionPath(product, subPath),
            this.artifactManager,
          ),
        weight: this.weight(subPath),
      }
    }

    if(UnknownProductType.isUnknownProductType(toType))
      return {
        fromType,
        toType,
        convert: async (product: Product) =>
          UnknownProduct.create(product),
        weight: 0,
      }

    if(
      MarkedProductType.isMarkedProductType(fromType)
      && MarkedProductType.isMarkedProductType(toType)
      && fromType.marker === toType.marker
    ) {
      const subPath = await this.compose(fromType.child, toType.child)
      return subPath && {
        fromType,
        toType,
        convert: async (product: MarkedProduct) =>
          MarkedProduct.create(product.marker, await this.executeConversionPath(product.child, subPath)),
        weight: this.weight(subPath),
      }
    }

    if(
      TupleProductType.isTupleProductType(fromType)
      && ArrayProductType.isArrayProductType(toType)
    )
      return {
        fromType: TupleProductType.create(Array(fromType.elementTypes.length).fill(toType.elementType)),
        toType,
        convert: async (product: TupleProduct) =>
          ArrayProduct.create(product.children),
        weight: 0,
      }

    if(
      ArrayProductType.isArrayProductType(fromType)
      && ArrayProductType.isArrayProductType(toType)
    ) {
      const subPath = await this.compose(fromType.elementType, toType.elementType)
      return subPath && {
        fromType,
        toType,
        convert: async (product: ArrayProduct) =>
          ArrayProduct.create(await Promise.all(product.children.map(child =>
            this.executeConversionPath(child, subPath),
          ))),
        weight: this.weight(subPath),
      }
    }

    if(
      TupleProductType.isTupleProductType(fromType)
      && TupleProductType.isTupleProductType(toType)
      && fromType.elementTypes.length === toType.elementTypes.length
    ) {
      const subPaths = await Promise.all(fromType.elementTypes.map((subFromType, i) =>
        this.compose(subFromType, toType.elementTypes[i]),
      ))
      if(subPaths.some(subPath => subPath === null))
        return null
      return {
        fromType,
        toType,
        convert: async (product: TupleProduct<any[]>) =>
          TupleProduct.create(await Promise.all(product.children.map((child, i) =>
            this.executeConversionPath(child, subPaths[i]!),
          ))),
        weight: subPaths.map(subPath => this.weight(subPath!)).reduce((a, b) => a + b),
      }
    }

    return null
  }

  private weight(path: ConversionPath){
    return path.reduce((a, b) => a + b.weight, 0)
  }

  private async executeConversionPath(product: Product, conversions: ConversionPath){
    if(!conversions.length) return product

    return await this.artifactManager.computeRef(
      [ConversionRegistry.artifactStoreId, conversions[conversions.length - 1].toType, product],
      async () => {
        let currentProduct = product
        for(const conversion of conversions)
          currentProduct = await artifactManager.computeRef(
            [ConversionRegistry.artifactStoreId, conversion.toType, currentProduct],
            async () => conversion.convert(currentProduct),
            this.excludeStores,
          )
        return currentProduct
      },
      this.excludeStores,
    )
  }

  async convertProduct<T extends Product, F extends ConvertibleTo<T>>(
    toType: ProductType<T>,
    from: F,
  ): Promise<T>{
    const fromType = Product.getProductType(from)

    const conversions = await this.compose(fromType, toType)

    if(!conversions)
      throw new Error(`Could not find path to convert product type ${Hash.create(fromType)} to ${Hash.create(toType)}`)

    return await this.executeConversionPath(from, conversions) as T
  }

  private rehydrateConversionPath(path: Array<Omit<ConversionImpl<any, any>, "convert">>): ConversionPath{
    return path.map(part => this.rehydrateConversion(part))
  }

  private rehydrateConversionMemo = new WeakMap<Omit<ConversionImpl<any, any>, "convert">, ConversionImpl<any, any>>()
  private rehydrateConversion(conversion: Omit<ConversionImpl<any, any>, "convert">): ConversionImpl<any, any>{
    if("convert" in conversion && typeof conversion["convert"] === "function")
      return conversion
    const existing = this.rehydrateConversionMemo.get(conversion)
    if(existing) return existing
    const rehydrated: ConversionImpl<any, any> = {
      ...conversion,
      // This function will get replaced when called
      convert: async (product: Product): Promise<Product> => {
        const { fromType, toType, id } = rehydrated
        if(id) {
          const registered = this.registered.get(id)
          if(!registered) throw new Error(`Could not find conversion with id ${id}`)
          Object.assign(rehydrated, registered)
        }
        else {
          const computed = await this.finishPathSegment(rehydrated.fromType, rehydrated.toType)
          if(!computed || !Hash.equal(fromType, computed.fromType) || !Hash.equal(toType, computed.toType))
            throw new Error(`Could not create conversion from ${Hash.create(fromType)} to ${Hash.create(toType)}`)
          Object.assign(rehydrated, computed)
        }
        if(rehydrated.convert === origConvert)
          throw new Error("Could not rehydrate conversion part")
        return rehydrated.convert(product)
      },
    }
    const origConvert = rehydrated.convert
    this.rehydrateConversionMemo.set(conversion, rehydrated)
    return rehydrated
  }

}

export const conversionRegistry = new ConversionRegistry(artifactManager)
