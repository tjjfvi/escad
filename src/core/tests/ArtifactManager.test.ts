
import { ArtifactManager, ArtifactStore, Hash, Id, WrappedValue } from "../src"

const createArtifactStoreMock = (
  name: string,
  output: unknown[],
  returnValues: (WrappedValue<unknown> | null)[],
): ArtifactStore => ({
  lookupRaw: async hash => {
    output.push([name, "lookupRaw", hash])
    return returnValues.pop() ?? null
  },
  lookupRef: async loc => {
    output.push([name, "lookupRef", loc])
    return returnValues.pop() ?? null
  },
  storeRaw: async (hash, value) => {
    output.push([name, "storeRaw", hash, value])
  },
  storeRef: async (loc, hash) => {
    output.push([name, "storeRef", loc, hash])
  },
})

test("", async () => {
  const artifactManager = new ArtifactManager()
  const output = [] as unknown[]
  const store0 = createArtifactStoreMock("store0", output, [])
  const store1 = createArtifactStoreMock("store1", output, [
    WrappedValue.create({ test0: true }),
    WrappedValue.create({ test1: true }),
    WrappedValue.create({ test2: true }),
    WrappedValue.create({ test3: true }),
  ])
  const excludeStores = new Set([store1])
  artifactManager.artifactStores.push({}, store0, store1)
  const artifacts = [
    { isTestArtifact: true },
    Hash.create("test"),
    Id.create(__filename, "@escad/core", "Test", "ArtifactTest"),
  ]
  for(const artifact of artifacts) {
    await artifactManager.lookupRaw(Hash.create(artifact))
    await artifactManager.lookupRaw(Hash.create(artifact), excludeStores)
    await artifactManager.storeRaw(artifact)
    await artifactManager.storeRaw(Promise.resolve(artifact), excludeStores)
  }
  await artifactManager.lookupRef(artifacts)
  await artifactManager.lookupRef(artifacts, excludeStores)
  await artifactManager.storeRef(artifacts, artifacts[0])
  await artifactManager.storeRef([], artifacts[1], excludeStores)
  expect(output).toMatchSnapshot()
})
