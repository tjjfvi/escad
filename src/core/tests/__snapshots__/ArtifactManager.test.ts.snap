export const snapshot = {};

snapshot[`ArtifactManager 1`] = `
[
  [
    "store0",
    "lookupRaw",
    "9057425f657d25f62d1adaf518df899b1703b0694e6f3e0a2ea64457be49b90f",
  ],
  [
    "store1",
    "lookupRaw",
    "9057425f657d25f62d1adaf518df899b1703b0694e6f3e0a2ea64457be49b90f",
  ],
  {
    isTestArtifact: true,
  },
  [
    "store0",
    "lookupRaw",
    "9057425f657d25f62d1adaf518df899b1703b0694e6f3e0a2ea64457be49b90f",
  ],
  null,
  [
    "store0",
    "storeRaw",
    "9057425f657d25f62d1adaf518df899b1703b0694e6f3e0a2ea64457be49b90f",
    {
      value: {
        isTestArtifact: true,
      },
    },
  ],
  [
    "store1",
    "storeRaw",
    "9057425f657d25f62d1adaf518df899b1703b0694e6f3e0a2ea64457be49b90f",
    {
      value: {
        isTestArtifact: true,
      },
    },
  ],
  [
    "store0",
    "storeRaw",
    "9057425f657d25f62d1adaf518df899b1703b0694e6f3e0a2ea64457be49b90f",
    {
      value: {
        isTestArtifact: true,
      },
    },
  ],
  [
    "store0",
    "lookupRaw",
    "6e9c3da9ea693b2a19048fc7312e0c877018aa0ee5195867051372f2e58dadd0",
  ],
  "0a843fe2c3373dd9f5072dc37674ccab20eb807c564ca34b6384e4cae0f318c2",
  [
    "store0",
    "lookupRaw",
    "6e9c3da9ea693b2a19048fc7312e0c877018aa0ee5195867051372f2e58dadd0",
  ],
  null,
  [
    "store0",
    "storeRaw",
    "6e9c3da9ea693b2a19048fc7312e0c877018aa0ee5195867051372f2e58dadd0",
    {
      value: "0a843fe2c3373dd9f5072dc37674ccab20eb807c564ca34b6384e4cae0f318c2",
    },
  ],
  [
    "store1",
    "storeRaw",
    "6e9c3da9ea693b2a19048fc7312e0c877018aa0ee5195867051372f2e58dadd0",
    {
      value: "0a843fe2c3373dd9f5072dc37674ccab20eb807c564ca34b6384e4cae0f318c2",
    },
  ],
  [
    "store0",
    "storeRaw",
    "6e9c3da9ea693b2a19048fc7312e0c877018aa0ee5195867051372f2e58dadd0",
    {
      value: "0a843fe2c3373dd9f5072dc37674ccab20eb807c564ca34b6384e4cae0f318c2",
    },
  ],
  [
    "store0",
    "lookupRaw",
    "fbbc3210f76789ad2ef86aec42dc8b3a296acd023029768e82ae7704e2ec28a6",
  ],
  [
    "store1",
    "lookupRaw",
    "fbbc3210f76789ad2ef86aec42dc8b3a296acd023029768e82ae7704e2ec28a6",
  ],
  null,
  [
    "store0",
    "lookupRaw",
    "fbbc3210f76789ad2ef86aec42dc8b3a296acd023029768e82ae7704e2ec28a6",
  ],
  null,
  [
    "store0",
    "storeRaw",
    "fbbc3210f76789ad2ef86aec42dc8b3a296acd023029768e82ae7704e2ec28a6",
    {
      value: "@escad/core/Test/ArtifactTest",
    },
  ],
  [
    "store1",
    "storeRaw",
    "fbbc3210f76789ad2ef86aec42dc8b3a296acd023029768e82ae7704e2ec28a6",
    {
      value: "@escad/core/Test/ArtifactTest",
    },
  ],
  [
    "store0",
    "storeRaw",
    "fbbc3210f76789ad2ef86aec42dc8b3a296acd023029768e82ae7704e2ec28a6",
    {
      value: "@escad/core/Test/ArtifactTest",
    },
  ],
  [
    "store0",
    "lookupRef",
    [
      {
        isTestArtifact: true,
      },
      "0a843fe2c3373dd9f5072dc37674ccab20eb807c564ca34b6384e4cae0f318c2",
      "@escad/core/Test/ArtifactTest",
    ],
  ],
  [
    "store1",
    "lookupRef",
    [
      {
        isTestArtifact: true,
      },
      "0a843fe2c3373dd9f5072dc37674ccab20eb807c564ca34b6384e4cae0f318c2",
      "@escad/core/Test/ArtifactTest",
    ],
  ],
  [
    "store0",
    "storeRaw",
    "08e4add5f841b3cae74852ca21866da144aba5a68d0c3eb329e53a7887820060",
    {
      value: {
        test2: true,
      },
    },
  ],
  [
    "store1",
    "storeRaw",
    "08e4add5f841b3cae74852ca21866da144aba5a68d0c3eb329e53a7887820060",
    {
      value: {
        test2: true,
      },
    },
  ],
  [
    "store0",
    "lookupRef",
    [
      {
        isTestArtifact: true,
      },
      "0a843fe2c3373dd9f5072dc37674ccab20eb807c564ca34b6384e4cae0f318c2",
      "@escad/core/Test/ArtifactTest",
    ],
  ],
  [
    "store0",
    "storeRef",
    [
      {
        isTestArtifact: true,
      },
      "0a843fe2c3373dd9f5072dc37674ccab20eb807c564ca34b6384e4cae0f318c2",
      "@escad/core/Test/ArtifactTest",
    ],
    "9057425f657d25f62d1adaf518df899b1703b0694e6f3e0a2ea64457be49b90f",
  ],
  [
    "store1",
    "storeRef",
    [
      {
        isTestArtifact: true,
      },
      "0a843fe2c3373dd9f5072dc37674ccab20eb807c564ca34b6384e4cae0f318c2",
      "@escad/core/Test/ArtifactTest",
    ],
    "9057425f657d25f62d1adaf518df899b1703b0694e6f3e0a2ea64457be49b90f",
  ],
  [
    "store0",
    "storeRaw",
    "9057425f657d25f62d1adaf518df899b1703b0694e6f3e0a2ea64457be49b90f",
    {
      value: {
        isTestArtifact: true,
      },
    },
  ],
  [
    "store1",
    "storeRaw",
    "9057425f657d25f62d1adaf518df899b1703b0694e6f3e0a2ea64457be49b90f",
    {
      value: {
        isTestArtifact: true,
      },
    },
  ],
  [
    "store0",
    "storeRaw",
    "6e9c3da9ea693b2a19048fc7312e0c877018aa0ee5195867051372f2e58dadd0",
    {
      value: "0a843fe2c3373dd9f5072dc37674ccab20eb807c564ca34b6384e4cae0f318c2",
    },
  ],
  [
    "store1",
    "storeRaw",
    "6e9c3da9ea693b2a19048fc7312e0c877018aa0ee5195867051372f2e58dadd0",
    {
      value: "0a843fe2c3373dd9f5072dc37674ccab20eb807c564ca34b6384e4cae0f318c2",
    },
  ],
  [
    "store0",
    "storeRaw",
    "fbbc3210f76789ad2ef86aec42dc8b3a296acd023029768e82ae7704e2ec28a6",
    {
      value: "@escad/core/Test/ArtifactTest",
    },
  ],
  [
    "store1",
    "storeRaw",
    "fbbc3210f76789ad2ef86aec42dc8b3a296acd023029768e82ae7704e2ec28a6",
    {
      value: "@escad/core/Test/ArtifactTest",
    },
  ],
  [
    "store0",
    "storeRef",
    [
    ],
    "6e9c3da9ea693b2a19048fc7312e0c877018aa0ee5195867051372f2e58dadd0",
  ],
  [
    "store0",
    "storeRaw",
    "6e9c3da9ea693b2a19048fc7312e0c877018aa0ee5195867051372f2e58dadd0",
    {
      value: "0a843fe2c3373dd9f5072dc37674ccab20eb807c564ca34b6384e4cae0f318c2",
    },
  ],
]
`;
